### ISSUE TITLE
Distributed training fails with criterion having `logging_outputs_can_be_summed` to `False`

### ISSUE THREAD

**Issue by mgaido91** [pytorch/fairseq#1840]
## 🐛 Bug

My understanding is that now criterions containing parameters are supported (as of https://github.com/pytorch/fairseq/commit/ba5f829f64d2adf9de4502dab94abf0a1a54106b).
I tried implementing a custom criterion with parameters and everything looked fine when testing on a single GPU.
Moving to a distributed training on 8 GPUs the training got stuck at the very beginning and never went on.

Debugging the issue, I noticed that all the processes started working on all the GPUs, rather than stick to the one they are assigned to.
Unfortunately, so far I have not been able to understand the root cause of the issue, though.

EDIT: further investigation showed that the problem is not related to the criterion having parameters but to the criterion having `logging_outputs_can_be_summed` returning `False`.

### To Reproduce

Add a very simple loss containing an attribute like:

```
@register_criterion("test_loss")
class TestLoss(FairseqCriterion):
    def __init__(self, args, task):
        super().__init__(args, task)
        self.real_criterion = TestLoss.build_real_criterion(args, task)

    @staticmethod
    def build_real_criterion(args, task):
        saved_criterion = args.criterion
        args.criterion = args.underlying_criterion
        assert saved_criterion != args.underlying_criterion
        underlying_criterion = task.build_criterion(args)
        args.criterion = saved_criterion
        return underlying_criterion

    @staticmethod
    def add_args(parser):
        parser.add_argument('--underlying-criterion', type=str, metavar='VAL', required=True,
                            help='underlying criterion to use for the model output loss')

    def forward(self, model, sample, reduce=True, log_probs=True):
        return self.real_criterion(model, sample, reduce=reduce)

    @staticmethod
    def reduce_metrics(logging_outputs):
        """Aggregate logging outputs from data parallel training."""
        loss_sum = utils.item(sum(log.get('loss', 0) for log in logging_outputs))
        sample_size = utils.item(sum(log.get('sample_size', 0) for log in logging_outputs))
        metrics.log_scalar('loss', loss_sum / sample_size / math.log(2), sample_size, round=3)
```

Train on multiple GPUs (in my case, 8).
The training gets stuck at 0% and never starts. The output of `nvidia-smi` contains:

```
+------------------------------------------------------------+
| Processes:                                      GPU Memory |
|  GPU       PID   Type   Process name            Usage      |
|============================================================|
|    0      2632      C   .../fairseq0.9/bin/python  1062MiB |
|    0      2633      C   .../fairseq0.9/bin/python   310MiB |
|    0      2634      C   .../fairseq0.9/bin/python   310MiB |
|    0      2635      C   .../fairseq0.9/bin/python   310MiB |
|    0      2636      C   .../fairseq0.9/bin/python   310MiB |
|    0      2637      C   .../fairseq0.9/bin/python   310MiB |
|    0      2638      C   .../fairseq0.9/bin/python   310MiB |
|    0      2639      C   .../fairseq0.9/bin/python   310MiB |
|    1      2632      C   .../fairseq0.9/bin/python   310MiB |
|    1      2633      C   .../fairseq0.9/bin/python  1016MiB |
|    1      2634      C   .../fairseq0.9/bin/python   310MiB |
|    1      2635      C   .../fairseq0.9/bin/python   310MiB |
|    1      2636      C   .../fairseq0.9/bin/python   310MiB |
|    1      2637      C   .../fairseq0.9/bin/python   310MiB |
|    1      2638      C   .../fairseq0.9/bin/python   310MiB |
|    1      2639      C   .../fairseq0.9/bin/python   310MiB |
|    2      2632      C   .../fairseq0.9/bin/python   310MiB |
|    2      2633      C   .../fairseq0.9/bin/python   310MiB |
|    2      2634      C   .../fairseq0.9/bin/python  1046MiB |
|    2      2635      C   .../fairseq0.9/bin/python   310MiB |
|    2      2636      C   .../fairseq0.9/bin/python   310MiB |
|    2      2637      C   .../fairseq0.9/bin/python   310MiB |
|    2      2638      C   .../fairseq0.9/bin/python   310MiB |
|    2      2639      C   .../fairseq0.9/bin/python   310MiB |
|    3      2632      C   .../fairseq0.9/bin/python   310MiB |
|    3      2633      C   .../fairseq0.9/bin/python   310MiB |
|    3      2634      C   .../fairseq0.9/bin/python   310MiB |
|    3      2635      C   .../fairseq0.9/bin/python  1048MiB |
|    3      2636      C   .../fairseq0.9/bin/python   310MiB |
|    3      2637      C   .../fairseq0.9/bin/python   310MiB |
|    3      2638      C   .../fairseq0.9/bin/python   310MiB |
|    3      2639      C   .../fairseq0.9/bin/python   310MiB |
|    4      2632      C   .../fairseq0.9/bin/python   310MiB |
|    4      2633      C   .../fairseq0.9/bin/python   310MiB |
|    4      2634      C   .../fairseq0.9/bin/python   310MiB |
|    4      2635      C   .../fairseq0.9/bin/python   310MiB |
|    4      2636      C   .../fairseq0.9/bin/python  1094MiB |
|    4      2637      C   .../fairseq0.9/bin/python   310MiB |
|    4      2638      C   .../fairseq0.9/bin/python   310MiB |
|    4      2639      C   .../fairseq0.9/bin/python   310MiB |
|    5      2632      C   .../fairseq0.9/bin/python   310MiB |
|    5      2633      C   .../fairseq0.9/bin/python   310MiB |
|    5      2634      C   .../fairseq0.9/bin/python   310MiB |
|    5      2635      C   .../fairseq0.9/bin/python   310MiB |
|    5      2636      C   .../fairseq0.9/bin/python   310MiB |
|    5      2637      C   .../fairseq0.9/bin/python  1022MiB |
|    5      2638      C   .../fairseq0.9/bin/python   310MiB |
|    5      2639      C   .../fairseq0.9/bin/python   310MiB |
|    6      2632      C   .../fairseq0.9/bin/python   310MiB |
|    6      2633      C   .../fairseq0.9/bin/python   310MiB |
|    6      2634      C   .../fairseq0.9/bin/python   310MiB |
|    6      2635      C   .../fairseq0.9/bin/python   310MiB |
|    6      2636      C   .../fairseq0.9/bin/python   310MiB |
|    6      2637      C   .../fairseq0.9/bin/python   310MiB |
|    6      2638      C   .../fairseq0.9/bin/python  1064MiB |
|    6      2639      C   .../fairseq0.9/bin/python   310MiB |
|    7      2632      C   .../fairseq0.9/bin/python   310MiB |
|    7      2633      C   .../fairseq0.9/bin/python   280MiB |
|    7      2634      C   .../fairseq0.9/bin/python   277MiB |
|    7      2635      C   .../fairseq0.9/bin/python   293MiB |
|    7      2636      C   .../fairseq0.9/bin/python   266MiB |
|    7      2637      C   .../fairseq0.9/bin/python   270MiB |
|    7      2638      C   .../fairseq0.9/bin/python   310MiB |
|    7      2639      C   .../fairseq0.9/bin/python  1058MiB |
+------------------------------------------------------------+
```

I also tried changing the `--num-workers` to 0, changing `--ddp-backend` but the behavior didn't change.

#### Code sample

See section above.

### Expected behavior

The training running fine in a distributed environment too.

### Environment

 - fairseq Version (e.g., 1.0 or master): master (updated at 6a5181509aa1fa7d260985157e77211753da544b)
 - PyTorch Version (e.g., 1.0): 1.4
 - OS (e.g., Linux): ubuntu
 - How you installed fairseq (`pip`, source): source
 - Build command you used (if compiling from source): python setup.py install -e
 - Python version: 3.6
 - CUDA/cuDNN version: 10.0
 - GPU models and configuration: 8 K80
 - Any other relevant information:

### Additional context

The bug has been introduced in https://github.com/pytorch/fairseq/commit/4cae68037d4ded8fdc6d9c13edbfbbcb21e39f7f.

**Comment by mgaido91** [comment]
cc @jcai1 @okhonko sorry for pinging you, but I saw you worked on the commit I mentioned, so you may know more about this. Thanks.

**Comment by myleott** [comment]
Are you using a custom dataset? Usually this happens when you have dataloading code that touches the GPU (or uses pinned memory), causing PyTorch to initialize itself on the GPU in the subprocess.

Does it work if you use a standard criterion? What is the rest of your code/model doing?

**Comment by mgaido91** [comment]
@myleott thanks for your answer. It works if using any criterion without parameters (eg. `cross_entropy`), but using the same criterion as "underlying criterion" in the repro above causes the issue.

The rest is just a normal `translation` task with `transformer` model. The command I am executing is:

```
python train.py $data_bin \
    -t $lang -s en \
    --clip-norm 20  \
    --max-source-positions 8000 --max-target-positions 8000 \
    --max-tokens 4500 \
    --save-dir $save_dir \
    --max-epoch 100 \
    --lr 1e-3 \
    --min-lr 1e-08 \
    --lr-scheduler inverse_sqrt \
    --arch transformer \
    --dropout 0.3 \
    --warmup-updates 4000 --warmup-init-lr 1e-7 \
    --update-freq 4 \
    --optimizer adam --adam-betas '(0.9, 0.98)' \
    --task translation \
    --skip-invalid-size-inputs-valid-test \
    --sentence-avg \
    --user-dir $my_ext_dir --criterion test_loss--underlying-criterion cross_entropy
```

With this command and the `test_loss` mentioned in the issue description, the problem can be reproduced consistently. So the only custom part is the criterion (which is actually very similar to the `composite_loss`, so I expect that to be affected by this issue too). Hence, I am confident this is a bug, not a question about how to extend fairseq, but unfortunately, I have not yet been able to understand the root cause. If there is anything I can do to further investigate this, I am very open to any suggestion. Thanks.

**Comment by mgaido91** [comment]
Ok, so I managed to understand more about the problem. The issue is caused by https://github.com/pytorch/fairseq/commit/4cae68037d4ded8fdc6d9c13edbfbbcb21e39f7f.

The problem happens only with criterions which don't set `logging_outputs_can_be_summed` to `True` (the default is `False`).  Hence I will update the issue description as the problem is not with parameters but the difference was related to `logging_outputs_can_be_summed`.

I'll try and find a fix in the next days, but I am not very familiar with this part of code, so if anyone else comes up with a solution, then please go ahead with it as it may take some time for me and I am not sure I would come up with the cleanest possible solution.

I am switching back to label bug as I think it is evident that this is a bug. Thanks.

**Comment by mgaido91** [comment]
I see only now that there is already https://github.com/pytorch/fairseq/issues/1826 which is probably the same issue. I am closing this in favor of it. Thanks @myleott .

**Comment by raotnameh** [comment]
@mgaido91 Sorry for asking. But since you have a good understanding of criterions. 

In the criterion we return the loss, sample_size, logging_outputs. 

What is we return loss/sample_size, 1, logging_outputs only i.e., divide the loss by sample_size and just pass 1 in its place?

**Comment by mgaido91** [comment]
I think it is not a good idea, as losses are aggregated later among all gpus, so you should avoid that

**Comment by raotnameh** [comment]
@mgaido91  Why is it not a good idea? 

Lets say we have 2 gpus such that 4 and 3 training examples on each. 

why not take a mean and then pass instead of passing the loss as a sum?

**Comment by mgaido91** [comment]
because you should sum their losses and divide by 7 instead of dividing the first by 4 and the second by 3

**Comment by raotnameh** [comment]
@mgaido91 I request you to pelase share any reading material on why this is preferred over the other. 

Furthermore, if I do not take a mean before then the loss is rescaled heavily to 4 in the first iteration.
